### DISCUSSION TITLE
How are we achieving parallel downloading here?

### DISCUSSION THREAD

**Discussion by nikhildr22** [TCNOco/TcNo-TCHT#9] [Q&A] [answered]
Hi,

I was amazed by the downloading speed boost because of parallelization. 
Can somebody explain what are we using here to achieve this? It'd be great to have script that does just downloading in blazing speed.
Thanks

**Comment by TCNOco** [comment, marked as answer]
Nothing crazy, just multithreaded downloads.

[File-DownloadMethods.ps1](https://github.com/TCNOco/TcNo-TCHT/blob/main/PowerShell/Modules/File-DownloadMethods.psm1)
1. `Initialize-Aria2` - This installs aria2 either through Chocolatey or by downloading the aria2 distribution, and extracting the exe.
2. `Get-Aria2File` - Makes sure aria2 is installed, then tried to download a file with the following command: `aria2c --disable-ipv6 -x 8 -s 8 --continue --out="$OutputPath" "$Url" --console-log-level=error --download-result=hide`.

  `--disable-ipv6` fixes issues when using Tailscale and other VPN software... For some reason finds those connections and tries to download through them, causing downloads to fail randomly. `-x 8` (--max-connection-per-server) allows multithreading over 8 connections to the server, `-s 8` (--split) Download a file using N connections. `--out="$outputPath\$_"` Download to path. `"$Url/$_"` URL to download from. `-console-log-level=error --download-result=hide` declutters the console during a download.

3. If aria2c isn't in the path, it tries running the exe directly `& $aria2Path`.
4. Otherwise it defaults to the singlethreaded built-in downloader `Get-FileFromWeb`.

`Get-Aria2Files` is similar, it just takes an array in and downloads each file after the other `$files | ForEach-Object`

If you're looking for a program to speed up general downloads, look into jDownloader.

**Reply by nikhildr22** [reply]
Thanks a lot for taking the time to explain.
I tried jDownloader, it does the job for me. 
[for those who might come across this comment, you can head over to jDownloader's official site and download the installer, their git repo  does not seem to have a well-documented readme for installation]
